rust/strprintf/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+// Problem statement for `strprintf` crate: provide a way to interpolate printf-style format
+// strings using native Rust types. For example, it should be possible to format a string
+// "%i %.2f %x" using values 42u32, 3.1415f64, and 255u8, and get a `std::string::String` "42 3.14
+// ff".
+//
+// This is the same as strprintf module we already have in C++.
+//
+// The problem can be solved by wrapping `libc::snprintf`, which is what we do both here and in
+// C++. However, our experience with C++ showed that we should constrain what types can be
+// formatted. Otherwise, complex objects like `String` will be passed over FFI and lead to
+// unexpected results (e.g. garbage strings).
+//
+// To achieve that, we provide a `Printfable` trait that's implemented only for types that our
+// formatting macro accepts. Everything else will result in a compile-time error. See the docs in
+// `trait` module for more on that.
+
+extern crate libc;
+
+pub mod specifiers_iterator;
+pub mod traits;
+
+use std::ffi::{CStr, CString};
+use std::vec::Vec;
+use traits::*;
+
+/// Helper function to `fmt!`. **Use it only through that macro!**
+///
+/// Returns a formatted string, or the size of the buffer that's necessary to hold the formatted
+/// string.
+#[doc(hidden)]
+pub fn fmt_arg_with_buffer_size<T>(
+    format_cstring: &CStr,
+    arg_c_repr_holder: &T,
+    buf_size: usize,
+) -> Result<String, usize>
+where
+    T: CReprHolder,
+{
+    let mut buffer = Vec::<u8>::with_capacity(buf_size);
+    // Filling the vector with ones because CString::new() doesn't want any zeroes in there. The
+    // last byte is left unused, so that CString::new() can put a terminating zero byte there
+    // without triggering a re-allocation.
+    buffer.resize(buf_size - 1, 1);
```

Comment:
I think we don't actually need to initialize the buffer since we're not going to read from it before the libc::snprintf call. The "no null-bytes" rule is in place so you don't accidentally truncate strings.

Reply:
Yes, initialization is unnecessary; only the resize is. However, there is no way to make Vec bigger without initializing its newly created cells, so I have to put some value in there.

Resize is necessary because `CString::from_vec_unchecked()` [calls `Vec::into_boxed_slice()`](https://doc.rust-lang.org/src/std/ffi/c_str.rs.html#352-356), which [drops excess capacity](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_boxed_slice). If we don't resize, we'll pass CString an empty Vec, CString will create a string of one element (null byte), and we will pass that into `snprintf` but will tell it that there are `buf_size` bytes available. That will lead to memory corruption. Learned this the hard way :) Here's a Playground I used to explain the problem to someone else: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=80573245ed657ea1d4f2a0e719c19ec7.

The comment is outdated, too—I replaced `CString::new()` with an unsafe call. I'll go update that; thank you for bringing this up!

Reply:
I updated the comment. How do you like the new version?

Reply:
Sorry for the late reply. Yes, the new version is very clear, thanks for pointing that out. :)

Reply:
Okay, cool!

I'm having second thoughts about using `CString` for `buffer`—it appears that for this use case, I should've gone with a `Vec<c_char>` or something. I'm working on this now; might request a review of one more commit later :)